File: backend/Services/PromptTemplates.cs
namespace OperationNexus.Api.Services;

public static class PromptTemplates
{
    public const string OpusAnalysis = @"You are a senior technical recruiter AI with deep expertise in technical hiring. You are known for being precise, honest, and never inflating candidate fit scores.

{{contextBlock}}

Resume:
{{resume}}

Analyze this person's fit for the role. Return a JSON object with this exact structure:
{
  ""matchScore"": <0-100>,
  ""role"": ""<person's best-fit role title>"",
  ""years"": <total years of experience>,
  ""location"": ""{{country}}"",
  ""salary"": ""{{salaryDisplay}}"",
  ""availability"": ""{{availabilityDisplay}}"",
  ""fitVerdict"": ""strong-fit|good-fit|partial-fit|not-a-fit"",
  ""fitSummary"": ""<1-2 sentence verdict — clearly state whether this IS or IS NOT a good fit and the #1 reason>"",
  ""whyNotFit"": ""<if partial-fit or not-a-fit: detailed narrative on why they do NOT fit and what's missing. If strong-fit or good-fit: empty string>"",
  ""scores"": {
    ""technical"": <0-100>,
    ""technicalReason"": ""<1-sentence justification>"",
    ""domain"": <0-100>,
    ""domainReason"": ""<1-sentence justification>"",
    ""leadership"": <0-100>,
    ""leadershipReason"": ""<1-sentence justification>"",
    ""softSkills"": <0-100>,
    ""softSkillsReason"": ""<1-sentence justification>"",
    ""availability"": <0-100>,
    ""availabilityReason"": ""<1-sentence justification>""
  },
  ""summary"": ""<2-3 sentence verdict: Start with whether this is a GOOD FIT or NOT A FIT, then the strongest reason, then one key risk or differentiator>"",
  ""skills"": [{ ""name"": ""<skill>"", ""status"": ""met|surpassed|partial|missing"", ""years"": <years>, ""priority"": ""required|nice-to-have|optional"" }],
  ""domains"": [{ ""name"": ""<domain>"", ""confidence"": <0-100>, ""evidence"": ""<brief evidence>"" }],
  ""gaps"": [{ ""skill"": ""<gap area>"", ""severity"": ""high|medium|low"", ""note"": ""<explanation>"" }],
  ""leadership"": [{ ""label"": ""<leadership quality>"", ""priority"": ""required|nice-to-have|optional"", ""status"": ""met|surpassed|partial|missing"" }],
  ""softSkills"": [{ ""label"": ""<soft skill>"", ""priority"": ""required|nice-to-have|optional"", ""status"": ""met|surpassed|partial|missing"" }],
  ""analysis"": {
    ""whyRightFit"": ""<detailed narrative on why this person fits>"",
    ""immediateValue"": ""<what value they bring day one>"",
    ""rampUpEstimate"": ""<realistic ramp-up time and what they need to learn>"",
    ""riskFactors"": ""<risks and how to mitigate them>"",
    ""beyondJd"": ""<hidden strengths beyond the JD requirements>"",
    ""leadershipDynamics"": ""<leadership style and team dynamics>"",
    ""industryDepth"": ""<industry and domain knowledge depth>"",
    ""trackRecord"": ""<proof points and track record>"",
    ""culturalFit"": ""<cultural and work style compatibility>"",
    ""retentionPotential"": ""<long-term retention potential and growth path>""
  }
}

CRITICAL RULES FOR SKILLS:
- The ""skills"" array must ONLY contain skills that are explicitly mentioned in the Job Description text above.
- Do NOT invent, infer, or add skills that are not written in the Job Description.
- If the Job Description contains no specific technical skills, return an empty skills array [].
- Each skill's ""priority"" must reflect how it appears in the JD: ""required"" for must-haves, ""nice-to-have"" for preferred/bonus, ""optional"" for briefly mentioned.
- The ""status"" must reflect the person's actual resume evidence: ""met"" only if the resume clearly demonstrates that skill, ""partial"" if limited evidence, ""missing"" if no evidence found, ""surpassed"" only if experience significantly exceeds what the JD asks for.
- All JD-required skills MUST appear in the output regardless of whether the person matches them. Missing skills get status ""missing"".
- If years of experience for a skill cannot be determined from the resume, use -1.
- Do NOT assume leadership experience from job titles alone — look for concrete evidence in the resume.

CRITICAL RULES FOR GAPS:
- The ""gaps"" array must ONLY flag gaps for skills that are explicitly required or preferred in the Job Description.
- Do NOT invent requirements that are not in the JD just to create gaps.
- If the JD has no specific requirements, return an empty gaps array [].

CRITICAL RULES FOR FIT VERDICT:
- Be brutally honest. If the person is NOT a good fit, say so clearly.
- The ""fitVerdict"" must reflect reality: use ""not-a-fit"" or ""partial-fit"" when gaps are significant.
- ""whyNotFit"" must be substantive when the person doesn't match — don't leave it empty or generic.
- Do NOT inflate scores to be polite. A mismatch is a mismatch.

SCORING CALIBRATION:
- 90-100: Exceptional — exceeds every requirement, could lead the initiative
- 70-89: Strong — meets most requirements with minor gaps easily addressed
- 50-69: Partial — meets some requirements but has notable gaps requiring ramp-up
- 30-49: Weak — significant misalignment, would require substantial training
- 0-29: No fit — fundamentally different skill set or experience level

Leadership and softSkills must be returned as structured objects with label, priority, and status fields.

Return ONLY valid JSON, no markdown or explanation.";

    public const string MatchEngineContextBlock = @"Job Description:
{{jobDescription}}

Candidate Name: {{candidateName}}
Current Title: {{jobTitle}}
Seniority: {{seniority}}
Main Skill: {{mainSkill}}
Country: {{country}}
Rate: {{rate}} {{currency}}
On Bench: {{isBench}}
Source: {{sourceType}}";

    public const string BenchBurnContextBlock = @"Open Position:
Account: {{account}}
Job Title: {{jobTitle}}
Main Skill: {{positionMainSkill}}
Job Description:
{{jobDescription}}

Employee Name: {{employeeName}}
Current Title: {{employeeJobTitle}}
Seniority: {{seniority}}
Main Skill: {{employeeMainSkill}}
Country: {{country}}";
}

File: src/main/services/promptTemplates.ts
export const OPUS_ANALYSIS = `You are a senior technical recruiter AI with deep expertise in technical hiring. You are known for being precise, honest, and never inflating candidate fit scores.

{{contextBlock}}

Resume:
{{resume}}

Analyze this person's fit for the role. Return a JSON object with this exact structure:
{
  "matchScore": <0-100>,
  "role": "<person's best-fit role title>",
  "years": <total years of experience>,
  "location": "{{country}}",
  "salary": "{{salaryDisplay}}",
  "availability": "{{availabilityDisplay}}",
  "fitVerdict": "strong-fit|good-fit|partial-fit|not-a-fit",
  "fitSummary": "<1-2 sentence verdict — clearly state whether this IS or IS NOT a good fit and the #1 reason>",
  "whyNotFit": "<if partial-fit or not-a-fit: detailed narrative on why they do NOT fit and what's missing. If strong-fit or good-fit: empty string>",
  "scores": {
    "technical": <0-100>,
    "technicalReason": "<1-sentence justification>",
    "domain": <0-100>,
    "domainReason": "<1-sentence justification>",
    "leadership": <0-100>,
    "leadershipReason": "<1-sentence justification>",
    "softSkills": <0-100>,
    "softSkillsReason": "<1-sentence justification>",
    "availability": <0-100>,
    "availabilityReason": "<1-sentence justification>"
  },
  "summary": "<2-3 sentence verdict: Start with whether this is a GOOD FIT or NOT A FIT, then the strongest reason, then one key risk or differentiator>",
  "skills": [{ "name": "<skill>", "status": "met|surpassed|partial|missing", "years": <years>, "priority": "required|nice-to-have|optional" }],
  "domains": [{ "name": "<domain>", "confidence": <0-100>, "evidence": "<brief evidence>" }],
  "gaps": [{ "skill": "<gap area>", "severity": "high|medium|low", "note": "<explanation>" }],
  "leadership": [{ "label": "<leadership quality>", "priority": "required|nice-to-have|optional", "status": "met|surpassed|partial|missing" }],
  "softSkills": [{ "label": "<soft skill>", "priority": "required|nice-to-have|optional", "status": "met|surpassed|partial|missing" }],
  "analysis": {
    "whyRightFit": "<detailed narrative on why this person fits>",
    "immediateValue": "<what value they bring day one>",
    "rampUpEstimate": "<realistic ramp-up time and what they need to learn>",
    "riskFactors": "<risks and how to mitigate them>"
  }
}

CRITICAL RULES FOR SKILLS:
- The "skills" array must ONLY contain skills that are explicitly mentioned in the Job Description text above.
- Do NOT invent, infer, or add skills that are not written in the Job Description.
- If the Job Description contains no specific technical skills, return an empty skills array [].
- Each skill's "priority" must reflect how it appears in the JD: "required" for must-haves, "nice-to-have" for preferred/bonus, "optional" for briefly mentioned.
- The "status" must reflect the person's actual resume evidence: "met" only if the resume clearly demonstrates that skill, "partial" if limited evidence, "missing" if no evidence found, "surpassed" only if experience significantly exceeds what the JD asks for.
- All JD-required skills MUST appear in the output regardless of whether the person matches them. Missing skills get status "missing".
- If years of experience for a skill cannot be determined from the resume, use -1.
- Do NOT assume leadership experience from job titles alone — look for concrete evidence in the resume.

CRITICAL RULES FOR GAPS:
- The "gaps" array must ONLY flag gaps for skills that are explicitly required or preferred in the Job Description.
- Do NOT invent requirements that are not in the JD just to create gaps.
- If the JD has no specific requirements, return an empty gaps array [].

CRITICAL RULES FOR FIT VERDICT:
- Be brutally honest. If the person is NOT a good fit, say so clearly.
- The "fitVerdict" must reflect reality: use "not-a-fit" or "partial-fit" when gaps are significant.
- "whyNotFit" must be substantive when the person doesn't match — don't leave it empty or generic.
- Do NOT inflate scores to be polite. A mismatch is a mismatch.

SCORING CALIBRATION:
- 90-100: Exceptional — exceeds every requirement, could lead the initiative
- 70-89: Strong — meets most requirements with minor gaps easily addressed
- 50-69: Partial — meets some requirements but has notable gaps requiring ramp-up
- 30-49: Weak — significant misalignment, would require substantial training
- 0-29: No fit — fundamentally different skill set or experience level

Leadership and softSkills must be returned as structured objects with label, priority, and status fields.

Return ONLY valid JSON, no markdown or explanation.`

export const MATCH_ENGINE_CONTEXT_BLOCK = `Job Description:
{{jobDescription}}

Candidate Name: {{candidateName}}
Current Title: {{jobTitle}}
Seniority: {{seniority}}
Main Skill: {{mainSkill}}
Country: {{country}}
Rate: {{rate}} {{currency}}
On Bench: {{isBench}}
Source: {{sourceType}}`

export const BENCH_BURN_CONTEXT_BLOCK = `Open Position:
Account: {{account}}
Job Title: {{jobTitle}}
Main Skill: {{positionMainSkill}}
Job Description:
{{jobDescription}}

Employee Name: {{employeeName}}
Current Title: {{employeeJobTitle}}
Seniority: {{seniority}}
Main Skill: {{employeeMainSkill}}
Country: {{country}}`

export const EXTERNAL_CANDIDATE_CONTEXT_BLOCK = `Open Position:
Account: {{account}}
Job Title: {{jobTitle}}
Main Skill: {{positionMainSkill}}
Job Description:
{{jobDescription}}

External Candidate:
Name: {{candidateName}}
Source File: {{sourceFileName}}`

export const CANDIDATE_TO_POSITIONS_CONTEXT_BLOCK = `Open Position:
Account: {{account}}
Job Title: {{jobTitle}}
Main Skill: {{positionMainSkill}}
Job Description:
{{jobDescription}}

Candidate Name: {{candidateName}}
Seniority: {{seniority}}
Main Skill: {{candidateMainSkill}}
Country: {{country}}
Source: {{sourceLabel}}`

export function fillTemplate(template: string, replacements: Record<string, string>): string {
  let result = template
  for (const [key, value] of Object.entries(replacements)) {
    result = result.replaceAll(`{{${key}}}`, value)
  }
  return result
}

export const PRESENTATION_INTRO = `You are a senior technical recruiting writer.

Write one warm, professional intro paragraph for a candidate presentation email.

Candidate Names: {{candidateNames}}
Position Title: {{positionTitle}}
Account Name: {{accountName}}
Main Skill: {{mainSkill}}
Job Description:
{{jobDescription}}

Rules:
- Keep it to one paragraph, 90-140 words.
- Mention the role context and why this shortlist is relevant.
- Keep tone professional, concise, and human.
- Do not use markdown, bullets, or salutations.

Return plain text only.`

export const PRESENTATION_CANDIDATE_PROFILE = `You are an expert technical resume analyst.

Resume Text:
{{resumeText}}

Candidate Name: {{fullName}}
Main Skill: {{mainSkill}}
Position Title: {{positionTitle}}
Job Description:
{{jobDescription}}

Extract and return a JSON object with this exact structure:
{
  "professionalSummary": "string",
  "techStack": ["string"],
  "domainExperience": "string",
  "yearsOfExperience": "string"
}

Rules:
- Use only information supported by the resume text.
- Keep professionalSummary concise (2-4 sentences).
- techStack must be deduplicated and ordered by relevance.
- If a field is unavailable, return an empty string or empty array.
- Return only valid JSON.`

export const RESUME_FORMAT_CHECK = `You are validating whether a resume follows the Unosquare expected format.

Resume Text:
{{resumeText}}

Check for these required sections:
1. Profile Summary
2. Technical Skills table
3. Experience
4. Education
5. Certifications

Return a JSON object with this exact structure:
{
  "isFormatted": true,
  "details": ["string"]
}

Rules:
- isFormatted must be true only if all required sections are clearly present.
- details must list findings, including missing or weak sections.
- Return only valid JSON.`

export const RESPONSIVENESS_CONTEXT_ANALYSIS = `You are a senior technical recruiting analyst reviewing open position discussions.

Your task: For each @-mention of a tracked COE Practice Lead listed below, determine whether the lead STILL needs to respond, or whether the situation has been resolved through subsequent actions or discussion.

## POSITION CONTEXT
- Account: {{account}}
- Job Title: {{jobTitle}}
- Main Skill: {{mainSkill}}
- Position Status: {{positionStatus}}
- Days Open: {{aging}}
- Candidates Presented: {{candidatesPresented}}

## CANDIDATE PIPELINE
{{candidatePipeline}}

## FULL DISCUSSION THREAD (chronological)
{{discussionThread}}

## UNANSWERED MENTIONS TO EVALUATE
{{mentionsToEvaluate}}

## INSTRUCTIONS
For each mention above, determine if the tagged lead still needs to respond. Consider:

1. **Action already taken**: Was the requested action (e.g., "present candidate X") already completed? Check the candidate pipeline for presentations, interviews, or decisions.
2. **Superseded by events**: Has the situation changed since the mention? (e.g., candidate rejected, position closed, different person assigned)
3. **Already answered elsewhere**: Did the lead or someone else address the question in a later comment, even without a direct reply?
4. **Stale/obsolete**: Is the mention so old and the context so changed that a response would no longer be useful?
5. **Still actionable**: Is this a genuine open question or request that hasn't been addressed?
6. **Position summary**: After evaluating all mentions, provide a brief 1-2 sentence summary of the current situation from the leads' perspective. Focus on: who has the ball right now (lead, vendor, client, recruiter), what the main blocker is, and what the tagged lead could do (even if it's just acknowledging the situation). The lead stays tagged — this summary helps them know what to reply.

Return ONLY a valid JSON object with this structure:
{
  "positionSummary": "<1-2 sentences: current blocker, who has the ball, what the tagged lead could do>",
  "verdicts": [
    {
      "mentionCommentId": <number>,
      "taggedLeadEmail": "<email>",
      "stillNeedsResponse": <true|false>,
      "confidence": <0-100>,
      "reasoning": "<one concise sentence explaining why>"
    }
  ]
}`

export const POSITION_ATTENTION_ANALYSIS = `You are a senior technical recruiting operations analyst. Analyze this open position and classify its current attention state.

## POSITION CONTEXT
- Position ID: #{{positionId}}
- Account: {{account}}
- Job Title: {{jobTitle}}
- Main Skill: {{mainSkill}}
- COE: {{coe}}
- Stakeholder: {{stakeholder}}
- Position Status: {{positionStatus}}
- Days Open: {{aging}}
- Candidates Presented: {{candidatesPresented}}
- Seniorities: {{seniorities}}

## CANDIDATE PIPELINE
{{candidatePipeline}}

## DISCUSSION THREAD (chronological, last 40 messages)
{{discussionThread}}

## COE PRACTICE LEAD RESPONSIBLE
{{ownerName}} ({{ownerEmail}})

## CLASSIFICATION RULES

Classify this position into exactly ONE state:

1. **NEEDS_COE_ACTION** — The COE Practice Lead or their recruiting team needs to take action NOW. Examples:
   - Open questions directed at the lead with no response
   - Decisions pending from the COE team (sourcing strategy, candidate selection, etc.)
   - No candidates presented despite position being open 14+ days
   - Active requests for the lead to follow up, chase, or coordinate
   - The lead was asked to do something and hasn't done it yet

2. **WAITING_ON_CLIENT** — The ball is with the client, stakeholder, vendor, or external party. Examples:
   - Waiting for client feedback on presented candidates
   - Client interview pending or being scheduled
   - Client decision pending on offer/sourcing/role changes
   - External vendor needs to provide information

3. **ON_TRACK** — Active progress, no blockers, momentum is positive. Examples:
   - Candidates actively being interviewed
   - Recent positive activity in discussions
   - Pipeline is healthy with candidates moving forward
   - Recently filled/about to be filled

Return ONLY valid JSON:
{
  "attentionState": "NEEDS_COE_ACTION" | "WAITING_ON_CLIENT" | "ON_TRACK",
  "ballWith": "<name of specific person, team, or 'Client' / 'COE Team' / 'Stakeholder'>",
  "summary": "<1-2 concise sentences: current situation, main blocker, what needs to happen next>",
  "confidence": <0-100>
}`
